mediatek: mt7622: expand kernel size to 6MiB for Xiaomi AX6S (RB03/RB01)

remittor opened the pull request:
Expand kernel partition size on Xiaomi AX6S for the kernel larger than 4 MiB.
To prevent upgrading from old firmware before this commit, bump the compat version to 1.1

Manual upgrade from OpenWRT to new OpenWRT:
```
cd /tmp
dd if=factory.bin bs=1M count=4 | mtd write - kernel
dd if=factory.bin bs=1M skip=4 count=2 | mtd write - ubi
dd if=factory.bin bs=1M skip=6  | mtd -p 2097152 write - ubi
echo c > /proc/sysrq-trigger
```

OpenWRT install under Stock firmware:
1) XMiR-Patcher
2) https://openwrt.org/toh/xiaomi/ax3200#installation

rany2 commented:
If I upgrade using sysupgrade (on the factory.bin image) instead, what would happen? I think as a user if I didn't know that I needed to use those dd commands, I would probably "brick" it.

I think it's worth including those dd commands in that error message.

remittor commented:
> If I upgrade using sysupgrade (on the factory.bin image) instead, what would happen?

Without flag `-F`, the `sysupgrade` will generate an error.

> I think as a user if I didn't know that I needed to use those dd commands, I would probably "brick" it.

With `-F`, the `sysupgrade` will brick router.

musashino205 commented on target/linux/mediatek/dts/mt7622-xiaomi-redmi-router-ax6s.dts:
```diff
@@ -312,7 +312,7 @@
 			 */
 			partition@6c0000 {
```
This unit address should also be updated.

remittor replied:
fixed

981213 commented:
Hi! Thanks for your patch!
But as I said elsewhere, I'll be coming up with a different solution for this.

I'll mark this PR as blocked for now.

ptpt52 commented:
> Hi! Thanks for your patch! But as I said elsewhere, I'll be coming up with a different solution for this.
> 
> I'll mark this PR as blocked for now.

Allow me to guess, are you writing a loader to boot the kernel from a UBI volume?

remittor committed:
mediatek: mt7622: expand kernel size to 6MiB for Xiaomi AX6S (RB03/RB01)

Expand kernel partition size on Xiaomi AX6S for the kernel larger than
4 MiB.
To prevent upgrading from old firmware before this commit, bump the
compat version to 1.1

Signed-off-by: Oleg S <remittor@gmail.com>

remittor commented:
@ptpt52 
> Allow me to guess, are you writing a **loader to boot the kernel from a UBI volume**?

This implementation will be very useful.
It turns out that 4 MB will be allocated for a tiny loader. And the Linux kernel itself will take up an additional 4 MB in the UBII volume.

rany2 commented:
> It turns out that 4 MB will be allocated for a tiny loader. And the Linux kernel itself will take up an additional 4 MB in the UBI partition.

Why wouldn't the kernel partition be shrunk if it will only contain the tiny loader? It would possibly mean more free space in the UBI partition if the kernel is less than 4MB.

981213 commented:
> Why wouldn't the kernel partition be shrunk if it will only contain the tiny loader?

Actually I'm planning 512k for that.

rany2 commented:
> Actually I'm planning 512k for that.

That's great, I'm looking forward to learn how this would be pulled off because quite frankly this is all magic to me right now!

981213 commented:
See: #14770
It's made using a xiaomi flash dump on my mt7622 dev board, and it's not tested on any real hardware.
It would be great if anyone can help reviewing and testing it!

rany2 commented:
I've had more success with this PR than https://github.com/openwrt/openwrt/pull/14770 unfortunately, even though that other PR is more impressive and practical in the long-term (when it works properly). I'm sometimes having issues with the second u-boot constantly requesting that itb file without doing anything.

Anyway back to this PR, the original instructions should be amended though, before running those `dd` commands and resetting the machine; remount `/` and `/overlay` to `ro` (`mount -o remount,ro /` && `mount -o remount,ro /overlay`). Otherwise, it doesn't work consistently.

edrikk commented:
My vote goes to this PR due to the lack of a need to get to the routers/APs to update them.  eg if not running on 193.168.1.X and especially running as APs.

I appreciate the technical change of the other PR, but realistically, the Kernel will not out-grow this PR’s size before the device has to be replaced anyways.

ptpt52 commented:
it is possible to reduce the kernel size <= 4MB via disabling some config: `CONFIG_KERNEL_KALLSYMS CONFIG_KERNEL_DEBUG_INFO`

remittor commented:
@edrikk 
> My vote goes to this PR due to the lack of a need to get to the routers/APs to update them. eg if not running on 193.168.1.X and especially running as APs.

Even using the ubi-loader, you can do without using a local TFTP server.
You just need to ask the maintainers very nicely to add additional functionality to this ubi-loader.
This is exactly what I'm doing right now. Now I will study the code of this ubi-loader myself...

rany2 commented:
> it is possible to reduce the kernel size <= 4MB via disabling some config: `CONFIG_KERNEL_KALLSYMS CONFIG_KERNEL_DEBUG_INFO`

Wrong, bad idea. I tried out the 6.6 kernel and this doesn't work anymore.

rany2 commented:
> My vote goes to this PR due to the lack of a need to get to the routers/APs to update them. eg if not running on 193.168.1.X and especially running as APs.
> 
> I appreciate the technical change of the other PR, but realistically, the Kernel will not out-grow this PR’s size before the device has to be replaced anyways.

If the other PR is able to add support for factory.bin image, then I'd go with it. Until then, I agree with you. The current method of flashing is really finicky and unreliable, I'm getting strange issues with it constantly requesting that sysupgrade image on a loop via tftp ...

This solution is less impressive but works reliably (as long as you don't forget to make ubi read-only). Also, the compat versions aren't conflicting (1.1 vs 2.0). So it could be merged until that other PR could be fixed

rany2 commented:
It looks like https://github.com/openwrt/openwrt/pull/14770 now builds a factory.bin as well, if it works fine for me I guess we could just ignore this PR.

rany2 commented:
Seems like I still have the same issue with it even after flashing the factory.bin to firmware (from stock), I guess this PR might be preferred until the issues with the second u-boot approach is sorted out.

981213 commented:
#14770 is merged and this PR can be closed. :)
